File: algorithm/이것이코딩테스트다스터디/탐색/네트워크.js
// https://school.programmers.co.kr/learn/courses/30/lessons/43162
n = 2
computers = [[1, 1, 0], [1, 1, 1], [0, 1, 1]]
function solution(n, computers) {
    const arr = new Array(n+1);
    const ParentSet = new Set()
    for (let i = 1; i <= n; i++) {
        arr[i] = i;
    }
    console.log(arr)
    for (i =0; i<computers.length; i++)
    {
        for (j =0; j<computers[i].length; j++)
        {
            if (computers[i][j] == 1)
            {
                union(arr,i+1,j+1)
            }
        }
    }
    for (let i = 1; i <= n; i++) {
        ParentSet.add(getParent(arr, i))
    }
    console.log(ParentSet.size)
    return ParentSet.size;
}

 function getParent(arr, n) {
    if (arr[n] === n) return n;
  
    return (arr[n] = getParent(arr, arr[n]));
  }
  
  function union(arr, a, b) {
    a = getParent(arr, a);
    b = getParent(arr, b);
  
    if (a < b) arr[b] = a;
    else arr[a] = b;
  }
  
solution(n,computers)

File: Algorithm/python/programers/네트워크.py
# https://school.programmers.co.kr/learn/courses/30/lessons/43162

def solution(n, computers):
    parent = [i for i in range(n)]
    
    def find(v1):
        if v1 != parent[v1]:
            parent[v1] = find(parent[v1])
            return parent[v1]
        else:
            return v1
    
    def union(v1, v2):
        p1 = find(v1)
        p2 = find(v2)
        if p1 > p2:
            parent[p1] = p2
        else:
            parent[p2] = p1
            
    for i, computer in enumerate(computers):
        for j, is_connected in enumerate(computer):
            if i != j and is_connected == 1:
                p1 = find(i)
                p2 = find(j)
                if p1 != p2:
                    union(p1, p2)
                    
    for i in range(len(parent)):
        parent[i] = find(i)
    
    return len(set(parent))
